File: built-in plugins dsp/mb-util-delay-comp.cpp
/**
 * MB Delay Compensation
 * Category : effect
 * Type     : delay
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Manual sample-accurate delay for phase alignment
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_UTIL_DELAY_COMP_H
#define MB_UTIL_DELAY_COMP_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbUtilDelayComp : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-util-delay-comp";
    static constexpr const char* PLUGIN_NAME    = "MB Delay Compensation";
    static constexpr const char* PLUGIN_TYPE    = "delay";
    static constexpr const char* PLUGIN_CATEGORY = "effect";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float samples = 0f;  // range [0, 4096]
    float channel = 0f;  // range [0, 2]
    };

    MbUtilDelayComp() = default;
    ~MbUtilDelayComp() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.samples = std::clamp(params.samples, 0f, 4096f);
        params.channel = std::clamp(params.channel, 0f, 2f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Delay Compensation
        return input;
    }
};

#endif // MB_UTIL_DELAY_COMP_H

File: built-in plugins dsp/mb-util-delay-comp.ts
import type { PluginDefinition } from '../server/services/pluginHostService';

const MbUtilDelayCompPlugin: PluginDefinition = { id: 'mb-util-delay-comp', slug: 'mb-util-delay-comp', name: 'MB Delay Compensation', category: 'effect', type: 'delay' as any, version: '1.0.0', description: 'Manual sample-accurate delay for phase alignment', author: 'Max Booster', parameters: [{ id: 'samples', name: 'Delay Samples', type: 'float', defaultValue: 0, minValue: 0, maxValue: 4096, automatable: false }, { id: 'channel', name: 'Channel', type: 'float', defaultValue: 0, minValue: 0, maxValue: 2, automatable: false }], defaultPreset: { samples: 0, channel: 0 } };

export default MbUtilDelayCompPlugin;
